synthesizer/program/src/resources/credits.aleo

```diff
@@ -106,573 +149,496 @@ record credits:
+    cast true r3 into r4 as committee_state;
     // Retrieve the committee state of the specified validator.
-    get.or_use committee[r0] r9 into r10;
+    get.or_use committee[r0] r4 into r5;
+
+    // Fail if attempting to change commission. Commission can only be set when validator bonds the first time.
+    assert.eq r3 r5.commission;
 
-    // Increment the stake for the specified validator.
-    add r10.microcredits r3 into r11;
     // Construct the updated committee state.
-    cast r11 r10.is_open into r12 as committee_state;
+    cast r5.is_open r5.commission into r6 as committee_state;
 
     /* Bonded */
 
     // Construct the initial bond state.
-    cast r0 0u64 into r13 as bond_state;
+    cast r0 0u64 into r7 as bond_state;
     // Get the bond state for the caller, or default to the initial bond state.
-    get.or_use bonded[r0] r13 into r14;
+    get.or_use bonded[r0] r7 into r8;
     // Enforce the validator matches in the bond state.
-    assert.eq r14.validator r0;
-
+    assert.eq r8.validator r0;
+    
     // Increment the microcredits in the bond state.
-    add r14.microcredits r3 into r15;
+    add r8.microcredits r2 into r9;
+
+    // Include total pre-delegated amount when deciding if validator has enough credits to join committee.
+    // First get the current delegated amount, or start with 0 if none delegated.
+    get.or_use delegated[r0] 0u64 into r10;
+    // Self-bond amount + existing delegated amount = total bond.
+    add r2 r10 into r11;
     // Determine if the amount is at least 10 million credits.
-    gte r15 10_000_000_000_000u64 into r16;
+    gte r11 10_000_000_000_000u64 into r12;
     // Enforce the amount is at least 10 million credits.
-    assert.eq r16 true;
+    assert.eq r12 true;
 
     // Construct the updated bond state.
-    cast r0 r15 into r17 as bond_state;
+    cast r0 r9 into r13 as bond_state;
 
     /* Account */
 
     // Get the balance of the caller.
     // If the account does not exist, this finalize scope will fail.
-    get account[r0] into r18;
+    get account[r0] into r14;
     // Decrement the balance of the caller.
-    sub r18 r3 into r19;
+    sub r14 r2 into r15;
 
     /* Writes */
 
+    // if (new_validator)
+    contains committee[r0] into r16;
+    branch.eq r16 true to validator_in_committee;
+    // {
+        // Check if initial bond amount is at least 100 credits
+        gte r2 100_000_000u64 into r17;
+        // Assert that initial self bond is at least 100 credits
+        assert.eq r17 true;
+
+        // Get the committee size.
+        get.or_use metadata[aleo1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq3ljyzc] 0u32 into r18;
+        // Increment the committee size by one.
+        add r18 1u32 into r19;
+        // Set the new committee size.
+        set r19 into metadata[aleo1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq3ljyzc];
+        // Set the withdrawal address.
+        set r1 into withdraw[r0];
```

Comment:
Set should be done at top right after validating withdrawal address is not changing.